core/src/main/scala/com/softwaremill/kmq/redelivery/streams/RedeliveryTrackerStream.scala

```diff
@@ -1,37 +1,38 @@
 package com.softwaremill.kmq.redelivery.streams
 
 import akka.Done
-import akka.actor.ActorSystem
+import akka.actor.{ActorSystem, Cancellable}
 import akka.kafka.ConsumerMessage.CommittableMessage
 import akka.kafka.scaladsl.Consumer
-import akka.kafka.scaladsl.Consumer.DrainingControl
+import akka.kafka.scaladsl.Consumer.{Control, DrainingControl}
 import akka.kafka.{ConsumerSettings, Subscriptions}
-import akka.stream.ClosedShape
-import akka.stream.scaladsl.{Broadcast, Flow, GraphDSL, RunnableGraph, Sink}
+import akka.stream._
+import akka.stream.scaladsl.{Broadcast, Flow, GraphDSL, Keep, RunnableGraph, Sink}
+import akka.stream.stage.{GraphStageLogic, GraphStageWithMaterializedValue, InHandler, OutHandler}
 import com.softwaremill.kmq._
 import com.typesafe.scalalogging.StrictLogging
+import org.apache.kafka.common.{Metric, MetricName}
 
 import java.time.Clock
+import scala.collection.mutable
 import scala.concurrent.{ExecutionContext, Future}
 
 /**
  * Combines functionality of [[RedeliverySink]] and [[CommitMarkerSink]].
  */
 class RedeliveryTrackerStream(markerConsumerSettings: ConsumerSettings[MarkerKey, MarkerValue],
-                              markersTopic: String, maxPartitions: Int)
-                             (implicit system: ActorSystem, ec: ExecutionContext,
-                              kafkaClients: KafkaClients, kmqConfig: KmqConfig, clock: Clock) extends StrictLogging {
+                              kafkaClients: KafkaClients, kmqConfig: KmqConfig, maxPartitions: Int)
+                             (implicit system: ActorSystem, ec: ExecutionContext, clock: Clock) extends StrictLogging {
 
   def run(): DrainingControl[Done] = {
-    Consumer.committablePartitionedSource(markerConsumerSettings, Subscriptions.topics(markersTopic))
+    Consumer.committablePartitionedSource(markerConsumerSettings, Subscriptions.topics(kmqConfig.getMarkerTopic))
       .mapAsyncUnordered(maxPartitions) {
         case (topicPartition, source) =>
-
-          val redeliverySink = RedeliverySink(topicPartition.partition)
+          val redeliverySink = RedeliverySink(kafkaClients, kmqConfig)(topicPartition.partition)
           val commitMarkerSink = CommitMarkerSink()
 
-          RunnableGraph
-            .fromGraph(GraphDSL.createGraph(redeliverySink, commitMarkerSink)(combineFutures) {
+          val cancellable = RunnableGraph
+            .fromGraph(GraphDSL.createGraph(redeliverySink, commitMarkerSink)(Keep.left) {
```

Comment:
don't we need to cancel both sinks?

Reply:
What's special in the RedeliverySink is an additional TickSource that is instantiated there. Without the TickSource both sinks appear to be closed correctly and the DrainingControl is able to shut down the stream.

Here I'm exposing a hook to the TickSource as Cancellable materialized view from the whole RedeliverySink.